### ISSUE TITLE
Certain numbers (extremely large negative ones, at least) result in encoded binary updates that cannot be decoded

### ISSUE THREAD

**Issue by cb22** [yjs/yjs#411]
**Checklist**

* [X] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
* [X] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Describe the bug**
When trying to encode a really large number in a YArray, the resulting update is corrupted and cannot be decoded. I'm not sure if this belongs here, or in lib0 - so apologies if it's in the wrong place!

**To Reproduce**
Here's a test case that reproduces the issue repeatably:
```
const y = require('yjs')

// Seems to start somewhere around e+24. Doesn't happen for positive, at least in my quick tests.
const testData = [{"key": -2.2062063918362897e+50}]

const yDoc = new y.Doc()
const arr = yDoc.getArray('test')

arr.push(testData)

const update = y.encodeStateAsUpdateV2(yDoc)
const newYDoc = new y.Doc()

// This fails
y.applyUpdateV2(newYDoc, update)
```

**Expected behavior**
The update encodes and decodes as normal

**Environment Information**
 - This is in Node 14, but the same behaviour happens in a browser, and Node 17.
 - yjs@13.5.32

**Additional context**
We came across this from a service we have that converts data from an old encoding format used by an ancient version (I think 13.0.0-66?) - those state snapshots themselves *seemed* uncorrupted and handled the binary encoding fine, if that helps in any way.

**Comment by cb22** [comment]
Looks like the bug is in lib0's encoding of numbers:

The line [here](https://github.com/dmonad/lib0/blob/73006dbdcbf9eda57d2fbfa10c431ee3b5470adc/encoding.js#L453) only checks for overflow, not underflow. Extending this check to be `number.isInteger(data) && data <= binary.BITS31 && data >= -1 * binary.BITS31` appears to fix the problem and the round-trip encode / decode now works at first glance!

(I'm not sure if the negative case needs an additional -1 in its check or not, haven't had the chance to look in much more detail yet)

**Comment by dmonad** [comment]
Thank you for uncovering this bug! Especially for finding the source of the problem.

I fixed the issue, added tests, and release a new `lib0` package. 

The issue should be fixed if you bump `lib0` to `>=0.2.48` or `yjs` to the latest release (`npm update yjs`).
